wgsl/index.bs

```diff
@@ -4557,11 +4555,271 @@ TODO: *Stub*: Expression evaluation
+
+Control-flow is said to be uniform at a point in the program if all invocations execute in lockstep at that program point.
+An expression is said to be uniform if it is in uniform control-flow, and all invocations compute the same value for it.
+The location of a local variable is said to be uniform, if all invocations have the same value in that variable at every program point where it is live.
+
+### Uniformity analysis overview ### {#uniformity-overview}
+
+In this section we specify an analysis that verifies that functions which are only safe to call in uniform control flow (barriers and derivatives) are only called in such a context.
+
+<div class="note">Note: This analysis has the following desirable properties:
+      - Sound (meaning that it rejects every program that would break the uniformity requirements of builtins)
+      - Linear time complexity
+      - Refactoring a piece of code into a function cannot make it invalid if it was valid before
+      - If the analysis refuses a program, it provides a straightforward chain of implications that can be used to craft a good error message
+</div>
+
+The analysis analyzes each function in turn, verifying that there is a context where it is safe to call this function (otherwise it rejects the program), and computing metadata about the function to help analyze its callers in turn. This means that the call graph must first be built, and functions must be analyzed from the leaves (functions that call no function outside the standard library) upwards (toward the entry point), so that whenever we analyze a function we have already computed
+the metadata for all of its callees. There is no risk of being trapped in a cycle, as recurrence is forbidden in the language.
+
+### Analyzing the uniformity requirements of a function ### {#uniformity-function}
+
+Each function is analyzed in two phases.
+
+The first phase walks over the AST of the function, building a directed graph along the way.
+The second phase explores that graph, resulting in either rejecting the program, or computing the constraints on calling this function.
+
+Note: Each vertex in this graph corresponds either to the statement "The control-flow at a specific program point must be uniform" or to "The value of a specific expression must be uniform" or to "A specific variable must always hold a uniform value" or is one of the special vertices True and False.
+      Each (directed) edge corresponds to an implication.
+
+In more detail, here is the algorithm for analyzing a function:
+
+* Create some vertices called "True", "False", "CF_start", "CF_return" and if the function is non-void a vertex called "Value_return"
+* Create one vertex for each parameter of the function which we'll call "arg_i"
+* Walk over the AST of the function, adding vertices and edges to the graph following the rules of the next section.
+* Look at the *Behaviors* of the function body treated as a statement (see next section):
```

Comment:
For readability, we might want to consider separating out the "behaviors" analysis into its own section, and putting that section before the uniformity analysis section. The behaviors analysis is necessary, even without considering uniformity, because it allows us to prove that a non-void function returns in all of its codepaths, that there are no statements after a control-flow-ending statement, there are no "break" statements outside of a loop, etc.

Reply:
"treated as a statement" = "use the rules from the section on statements". I'll try to rephrase this.

Splitting the behaviors analysis is a good idea. In practice it is probably more efficient to do a single pass, but for the spec having several passes may well be clearer.